source/common/http/header_map_impl.h

```diff
@@ -230,121 +242,253 @@ class HeaderMapImpl : public virtual HeaderMap, NonCopyable {
+
+/**
+ * Concrete implementation of RequestHeaderMap which allows for variable custom registered inline
+ * headers.
+ */
+class RequestHeaderMapImpl final : public TypedHeaderMapImpl<RequestHeaderMap>,
+                                   public InlineStorage {
+public:
+  static std::unique_ptr<RequestHeaderMapImpl> create() {
+    return std::unique_ptr<RequestHeaderMapImpl>(new (inlineHeadersSize()) RequestHeaderMapImpl());
+  }
+
   INLINE_REQ_HEADERS(DEFINE_INLINE_HEADER_FUNCS)
   INLINE_REQ_RESP_HEADERS(DEFINE_INLINE_HEADER_FUNCS)
 
-protected:
-  // Explicit inline headers for the request header map.
-  // TODO(mattklein123): This is mostly copied between all of the concrete header map types.
-  // In a future change we can either get rid of O(1) headers completely, or it should be possible
-  // to statically register all O(1) headers and move to a single dynamically sized class where we
-  // we reference the O(1) headers in the table by an offset.
-  struct AllInlineHeaders {
-    AllInlineHeaders() { clear(); }
-    void clear() { memset(this, 0, sizeof(*this)); }
-
-    INLINE_REQ_HEADERS(DEFINE_INLINE_HEADER_STRUCT)
-    INLINE_REQ_RESP_HEADERS(DEFINE_INLINE_HEADER_STRUCT)
+private:
+  struct HeaderHandleValues {
+    INLINE_REQ_HEADERS(DEFINE_HEADER_HANDLE)
+    INLINE_REQ_RESP_HEADERS(DEFINE_HEADER_HANDLE)
   };
 
-  absl::optional<StaticLookupResponse> staticLookup(absl::string_view key) override {
-    return StaticLookupTable<RequestHeaderMapImpl>::lookup(*this, key);
-  }
-  void clearInline() override { inline_headers_.clear(); }
+  using HeaderHandles = ConstSingleton<HeaderHandleValues>;
 
-  AllInlineHeaders inline_headers_;
+  RequestHeaderMapImpl() { clearInline(); }
+  void clearInline() override { memset(inline_headers_, 0, inlineHeadersSize()); }
+  const HeaderEntryImpl* const* constInlineHeaders() const override { return inline_headers_; }
+  HeaderEntryImpl** inlineHeaders() override { return inline_headers_; }
 
-  friend class HeaderMapImpl;
+  HeaderEntryImpl* inline_headers_[];
 };
 
-class RequestTrailerMapImpl : public HeaderMapImpl, public RequestTrailerMap {};
+/**
+ * Concrete implementation of RequestTrailerMap which allows for variable custom registered inline
+ * headers.
+ */
+class RequestTrailerMapImpl final : public TypedHeaderMapImpl<RequestTrailerMap>,
+                                    public InlineStorage {
+public:
+  static std::unique_ptr<RequestTrailerMapImpl> create() {
+    return std::unique_ptr<RequestTrailerMapImpl>(new (inlineHeadersSize())
+                                                      RequestTrailerMapImpl());
+  }
+
+private:
+  RequestTrailerMapImpl() { clearInline(); }
+  void clearInline() override { memset(inline_headers_, 0, inlineHeadersSize()); }
+  const HeaderEntryImpl* const* constInlineHeaders() const override { return inline_headers_; }
+  HeaderEntryImpl** inlineHeaders() override { return inline_headers_; }
```

Comment:
it feels...unexpected to have override methods in a private section. Should these be protected rather than prviate? ctor can be private of course